In [311]:
import pandas as pd
import numpy as np
import requests
from bs4 import BeautifulSoup


In [312]:
starbucks_response = requests.get("https://worldpopulationreview.com/state-rankings/starbucks-stores-by-state")

In [313]:
starbucks_soup = BeautifulSoup(starbucks_response.content, "html.parser")

In [314]:
table = (starbucks_soup.find_all("table"))[0]
table

<table class="wpr-table"><thead class="table-head bg-wpr-table_header_bg relative z-40 px-0 py-0 text-xs uppercase tracking-wider"><tr><th class="datatable-th bg-wpr-table_header_bg text-wpr-table_header border-wpr-table_border sticky top-0 border-b bg-clip-padding px-3 py-2 text-left align-bottom leading-4 md:px-4 pin left-0 z-50 border-l-0"><div class="flex flex-col items-stretch justify-start"><div class="flex flex-row items-center cursor-pointer select-none">State</div></div></th><th class="datatable-th bg-wpr-table_header_bg text-wpr-table_header border-wpr-table_border sticky top-0 border-b bg-clip-padding px-3 py-2 text-left align-bottom leading-4 md:px-4 z-40"><div class="flex flex-col items-stretch justify-start"><div class="flex flex-row items-center cursor-pointer select-none">Starbucks Stores 2023<svg class="icon icon-tabler icon-tabler-chevron-down" fill="none" height="14" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" viewbox="0 0 24

In [315]:
rows = []

# iterate over all rows in the population table
for row in table.find_all("tr")[1:]:

  # State is found in <th>
  state = row.find("th").text

  # Number of locations is found in <td>
  
  locations_2023 = row.find_all("td")[0].get_text(strip = True)

  locations_2021 = row.find_all("td")[1].get_text(strip = True)

  locations_2024 = row.find_all("td")[2].get_text(strip = True)

  rows.append({
    "State": state,
    "Starbucks Stores 2023": locations_2023,
    "Starbucks Stores 2021": locations_2021,
    "Starbucks Stores 2024": locations_2024
  })


In [316]:
starbucks = pd.DataFrame(rows)
starbucks.head()

State Starbucks Stores 2023 Starbucks Stores 2021 Starbucks Stores 2024
0  California                 3,080                 2,959                 3,117
1       Texas                 1,346                 1,215                 1,409
2     Florida                   844                   786                   892
3  Washington                   741                   739                   736
4    New York                   692                   643                   715

In [317]:
dunkin_response = requests.get("https://worldpopulationreview.com/state-rankings/dunkin-donuts-by-state")
dunkin_soup = BeautifulSoup(dunkin_response.content, "html.parser")

In [318]:
table = (dunkin_soup.find_all("table"))[0]

In [319]:
rows = []

# iterate over all rows in the population table
for row in table.find_all("tr")[1:]:

  # State is found in <th>
  state = row.find("th").text

  # Number of locations is found in <td>
  
  locations_2024 = row.find_all("td")[0].get_text(strip = True)

  locations_2023 = row.find_all("td")[1].get_text(strip = True)

  rows.append({
    "State": state,
    "Dunkin Stores 2024": locations_2024,
    "Dunkin Stores 2023": locations_2023
  })

In [320]:
dunkin = pd.DataFrame(rows)
dunkin.head()

State Dunkin Stores 2024 Dunkin Stores 2023
0       New York              1,431              1,414
1  Massachusetts              1,042              1,068
2        Florida                909                883
3     New Jersey                872                866
4       Illinois                711                692

In [321]:
merged_data = pd.merge(starbucks, dunkin, on = "State")
merged_data.head()

State Starbucks Stores 2023 Starbucks Stores 2021 Starbucks Stores 2024 Dunkin Stores 2024 Dunkin Stores 2023
0  California                 3,080                 2,959                 3,117                143                134
1       Texas                 1,346                 1,215                 1,409                226                196
2     Florida                   844                   786                   892                909                883
3  Washington                   741                   739                   736                  0                 19
4    New York                   692                   643                   715              1,431              1,414

In [322]:
population_response = requests.get("https://simple.wikipedia.org/wiki/List_of_U.S._states_by_population")
population_soup = BeautifulSoup(population_response.content, "html.parser")

In [323]:
table = (population_soup.find_all("table"))[0]

In [324]:
rows = []

# iterate over all rows in the population table
for row in table.find_all("tr")[1:]:

  # State is found in <th>
  state = row.find_all("td")[2].get_text(strip = True)

  # Number of locations is found in the select column of <td>
  population = row.find_all("td")[3].get_text(strip = True)

  rows.append({
    "State": state,
    "Population": population
  })  

In [325]:
population = pd.DataFrame(rows)
population.head()

State  Population
0    California  39,538,223
1         Texas  30,145,505
2       Florida  21,538,187
3      New York  20,201,249
4  Pennsylvania  13,002,700

In [326]:
merged_data2 = pd.merge(merged_data,population, on="State")
merged_data2.head()

State Starbucks Stores 2023 Starbucks Stores 2021 Starbucks Stores 2024 Dunkin Stores 2024 Dunkin Stores 2023  Population
0  California                 3,080                 2,959                 3,117                143                134  39,538,223
1       Texas                 1,346                 1,215                 1,409                226                196  30,145,505
2     Florida                   844                   786                   892                909                883  21,538,187
3  Washington                   741                   739                   736                  0                 19   7,705,281
4    New York                   692                   643                   715              1,431              1,414  20,201,249

In [327]:
merged_data2["Starbucks_Stock"] = 96
merged_data2["Dunkin_Stock"] = 106

In [328]:
regions = pd.DataFrame({
    'State': [
        'Alabama', 'Alaska', 'Arizona', 'Arkansas', 'California', 'Colorado',
        'Connecticut', 'Delaware', 'Florida', 'Georgia', 'Hawaii', 'Idaho',
        'Illinois', 'Indiana', 'Iowa', 'Kansas', 'Kentucky', 'Louisiana',
        'Maine', 'Maryland', 'Massachusetts', 'Michigan', 'Minnesota',
        'Mississippi', 'Missouri', 'Montana', 'Nebraska', 'Nevada',
        'New Hampshire', 'New Jersey', 'New Mexico', 'New York',
        'North Carolina', 'North Dakota', 'Ohio', 'Oklahoma', 'Oregon',
        'Pennsylvania', 'Rhode Island', 'South Carolina', 'South Dakota',
        'Tennessee', 'Texas', 'Utah', 'Vermont', 'Virginia', 'Washington',
        'West Virginia', 'Wisconsin', 'Wyoming'
    ],
    'Region': [
        'South', 'West', 'West', 'South', 'West', 'West', 'Northeast', 
        'South', 'South', 'South', 'West', 'West', 'Midwest', 'Midwest', 
        'Midwest', 'Midwest', 'South', 'South', 'Northeast', 'South', 
        'Northeast', 'Midwest', 'Midwest', 'South', 'Midwest', 'West', 
        'Midwest', 'West', 'Northeast', 'Northeast', 'West', 'Northeast',
        'South', 'Midwest', 'Midwest', 'South', 'West', 'Northeast', 
        'Northeast', 'South', 'Midwest', 'South', 'South', 'West', 
        'Northeast', 'South', 'West', 'South', 'Midwest', 'West'
    ]
})

In [329]:
merged_data3 = pd.merge(merged_data2, regions, on='State')


In [330]:
merged_data3['Starbucks Stores 2023'] = merged_data3['Starbucks Stores 2023'].str.replace(',', '').astype(int)
merged_data3['Starbucks Stores 2021'] = merged_data3['Starbucks Stores 2021'].str.replace(',', '').astype(int)
merged_data3['Dunkin Stores 2024'] = merged_data3['Dunkin Stores 2024'].str.replace(',', '').astype(int)
merged_data3['Dunkin Stores 2023'] = merged_data3['Dunkin Stores 2023'].str.replace(',', '').astype(int)
merged_data3['Population'] = merged_data3['Population'].str.replace(',', '').astype(int)

In [331]:
merged_data3

State  Starbucks Stores 2023  Starbucks Stores 2021  ... Starbucks_Stock  Dunkin_Stock     Region
0       California                   3080                   2959  ...              96           106       West
1            Texas                   1346                   1215  ...              96           106      South
2          Florida                    844                    786  ...              96           106      South
3       Washington                    741                    739  ...              96           106       West
4         New York                    692                    643  ...              96           106  Northeast
5         Illinois                    677                    621  ...              96           106    Midwest
6          Arizona                    548                    535  ...              96           106       West
7         Colorado                    495                    497  ...              96           106       West
8             Ohio                    491                    446  ...              96           106    Midwest
9         Virginia                    489                    456  ...              96           106      South
10          Oregon                    359                    345  ...              96           106       West
11    Pennsylvania                    357                    406  ...              96           106  Northeast
12  North Carolina                    338                    422  ...              96           106      South
13         Georgia                    326                    379  ...              96           106      South
14        Michigan                    283                    310  ...              96           106    Midwest
15   Massachusetts                    273                    279  ...              96           106  Northeast
16      New Jersey                    261                    293  ...              96           106  Northeast
17        Maryland                    257                    274  ...              96           106      South
18          Nevada                    253                    269  ...              96           106       West
19         Indiana                    221                    265  ...              96           106    Midwest
20        Missouri                    188                    208  ...              96           106    Midwest
21       Minnesota                    184                    212  ...              96           106    Midwest
22       Tennessee                    180                    237  ...              96           106      South
23        Kentucky                    166                    131  ...              96           106      South
24       Wisconsin                    145                    176  ...              96           106    Midwest
25  South Carolina                    131                    207  ...              96           106      South
26     Connecticut                    123                    125  ...              96           106  Northeast
27            Utah                    101                    126  ...              96           106       West
28          Hawaii                     99                    105  ...              96           106       West
29          Kansas                     94                    120  ...              96           106    Midwest
30            Iowa                     89                    142  ...              96           106    Midwest
31         Alabama                     85                     99  ...              96           106      South
32       Louisiana                     84                     10  ...              96           106      South
33        Oklahoma                     79                     94  ...              96           106      South
34      New Mexico                     76                     92  ...              96           106       West
35           Idah

In [336]:
merged_data3[["State", "Starbucks Stores 2023", "Dunkin Stores 2023", "Population"]].groupby("State").sum()

Starbucks Stores 2023  Dunkin Stores 2023  Population
State                                                                
Alabama                            85                  59     5024279
Alaska                             49                   0      733391
Arizona                           548                 102     7151502
Arkansas                           55                   9     3011524
California                       3080                 134    39538223
Colorado                          495                  43     5773714
Connecticut                       123                 480     3605944
Delaware                           25                  66      989948
Florida                           844                 883    21538187
Georgia                           326                 265    10711908
Hawaii                             99                   6     1455271
Idaho                              67                   0     1839106
Illinois                          677                 692    12812508
Indiana                           221                 104     6785528
Iowa                               89                  34     3190369
Kansas                             94                  28     2937880
Kentucky                          166                  41     4505836
Louisiana                          84                  16     4657757
Maine                              30                 159     1362359
Maryland                          257                 303     6177224
Massachusetts                     273                1068     7029917
Michigan                          283                 105    10077331
Minnesota                         184                  34     5706494
Mississippi                        32                  12     2961279
Missouri                          188                  50     6154913
Montana                            32                   0     1084225
Nebraska                           58                  23     1961504
Nevada                            253                  38     3104614
New Hampshire                      29                 218     1377529
New Jersey                        261                 866     9288994
New Mexico                         76                  19     2117522
New York                          692                1414    20201249
North Carolina                    338                 206    10439388
North Dakota                       13                   0      779094
Ohio                              491                 245    11799448
Oklahoma                           79                  18     3959353
Oregon                            359                   0     4237256
Pennsylvania                      357                 629    13002700
Rhode Island                       27                 160     1097379
South Carolina                    131                 103     5118425
South Dakota                       25                   0      886667
Tennessee                         180                 123     6910840
Texas                            1346                 196    30145505
Utah                              101                   1     3271616
Vermont                             8                  49      643077
Virginia                          489                 230     8631393
Washington                        741                  19     7705281
West Virginia                      25                  27     1793716
Wisconsin                         145                  83     5893718
Wyoming                            23                   1      576851

In [ ]:
grouped_data = merged_data3[["Region", "Starbucks Stores", "Dunkin Stores", "Population"]].groupby("Region").sum()
grouped_data["People per Starbucks Store"] = (grouped_data["Population"] / grouped_data["Starbucks Stores"]).round(2)
grouped_data["People per Dunkin Store"] = (grouped_data["Population"] / grouped_data["Dunkin Stores"]).round(2)

grouped_data

Looking at this table, I can conclude there are the most Starbucks locations in the West region. There are also the least amount of people per Starbucks Store in the West region. I can also conclude there are the most Dunkin Donuts locations in the Northeast region, with the least amount of people per Dunkin Stores in the Northeast region. Starbucks Dominates the West, Midwest, and the South regions, whereas Dunkin Donuts is the King in the Northeast region. 

In [363]:
merged_data4 = merged_data3
# Replace 0s with Nan for States that don't have any Dunkin Stores
merged_data4["Dunkin Stores 2023"] = merged_data4["Dunkin Stores 2023"].replace(0, np.nan)
merged_data4["People per Starbucks Store"] = merged_data4["Population"] / merged_data4["Starbucks Stores 2023"]
merged_data4["People per Dunkin Store"] = merged_data4["Population"] / merged_data4["Dunkin Stores 2023"]
merged_data4["Starbucks Stock per Store"] = merged_data4["Starbucks_Stock"] / merged_data4["Starbucks Stores 2023"]
merged_data4["Dunkin Stock per Store"] = merged_data4["Dunkin_Stock"] / merged_data4["Dunkin Stores 2023"]

merged_data4[["Region", "Starbucks Stock per Store", "Dunkin Stock per Store", "People per Starbucks Store", "People per Dunkin Store"]].groupby("Region").mean().round(2)


Starbucks Stock per Store  Dunkin Stock per Store  People per Starbucks Store  People per Dunkin Store
Region                                                                                                           
Midwest                         1.48                    2.06                    34165.44                 87388.27
Northeast                       2.66                    0.52                    41134.80                 10516.61
South                           1.19                    2.96                    42579.66                112627.38
West                            1.15                   27.55                    19104.39                576572.08

Looking at the table, Starbucks has the highest stock value per store in the Northeast region ($2.66) and the lowest in the West ($1.15). Dunkin' stock per store is highest in the West ($27.55) due to the lower prevalence of stores in that region, and lowest in the Northeast ($0.52), where it's most common. People per store also shows significant differences, with the West having the fewest people per Starbucks store (19,104) and the most per Dunkin' store (576,572), highlighting the stark difference in the distribution of Dunkin' stores across regions.
Starbucks has a more evenly distributed number of stores, with fewer people per store, especially in the West. Dunkin' is much more concentrated in the Northeast, where the people-per-store ratio is much lower than in other regions.


In [368]:
def webscrape(link):

    import requests
    from bs4 import BeautifulSoup

    response = requests.get(link)
    soup = BeautifulSoup(response.content, "html.parser")

    # Find the table in the website
    table=(soup.find_all("table"))[0]
    rows= []

# Iterate over all rows in the table
    for row in table.find_all("tr")[1:]:

        # State is found in <th>
        state = row.find("th").text

        # Number of locations is found in <td>
        locations = row.find("td").text

        rows.append({
                "State": state,
                "Stores": locations,
    })
    df = pd.DataFrame(rows)
    return df    

In [370]:
webscrape("https://worldpopulationreview.com/state-rankings/walmart-stores-by-state")

State Stores
0                  Texas    517
1                Florida    341
2             California    280
3         North Carolina    192
4                Georgia    189
5               Illinois    161
6                   Ohio    146
7               Missouri    137
8              Tennessee    137
9           Pennsylvania    134
10              Virginia    134
11               Alabama    131
12              Oklahoma    122
13              Arkansas    120
14             Louisiana    116
15               Indiana    113
16               Arizona    112
17        South Carolina    110
18              New York     98
19              Michigan     94
20              Kentucky     93
21             Wisconsin     89
22              Colorado     88
23           Mississippi     78
24                Kansas     74
25             Minnesota     68
26            Washington     66
27            New Jersey     62
28                  Iowa     60
29                  Utah     53
30              Maryland     48
31         Massachusetts     48
32            New Mexico     46
33                Oregon     45
34                Nevada     43
35              Nebraska     42
36         West Virginia     39
37           Connecticut     33
38                 Idaho     26
39         New Hampshire     26
40                 Maine     22
41          South Dakota     15
42               Montana     14
43          North Dakota     14
44               Wyoming     12
45                Hawaii     10
46                Alaska      9
47          Rhode Island      9
48               Vermont      6
49  District of Columbia      3